rfcs/20200306-single-client-parameter-server.md

```diff
@@ -0,0 +1,517 @@
+
+
+### Low-level Primitives
+
+We can potentially expose them in the future when they are more stable and when we want to allow more advanced use cases.
+
+We will have `Cluster` and `Worker` classes to encapsulate logic related to remote function scheduling.
+
+
+```python
+class Cluster(object):
+
+  def __init__(self, cluster_resolver, failure_handler=None):
+    """Create the cluster instance and connect to the remote cluster."""
+    pass
+
+  @property
+  def workers(self):
+    """Return all available workers."""
+    return self._workers
+
+  def schedule(self, function, args=None, kwargs=None):
+    """Schedule the function on one worker.
+
+    It adds the function to the global scheduling queue and returns future
+    objects immediately.
+    """
+    pass
+
+  def join(self):
+    """Block until all scheduled functions are complete."""
+    pass
+```
+
+
+We will probably merge this `Worker` with executors.
+
+
+```python
+class Worker(object):
+
+  def __init__(self,
+               worker_job_name,
+               cluster,
+               max_scheduled_functions=100):
+    """Create a scheduling queue and a thread that processes the queue."""
+    pass
+
+  def schedule(self, function, args=None, kwargs=None):
+    """Schedule the function on the worker.
+
+    It adds the function to the scheduling queue. It returns Future object
+    immediately unless the scheduling queue is full.
+    """
+    pass
+
+  def healthy(self):
+    """Return a boolean indicating whether the worker is health or not."""
+    pass
+
+  def _set_dead(self):
+    """Declare the worker is dead and poison all future objects."""
+    pass
+
+  def _rebuild_resources(self):
+    """Rebuild worker-local resources when it is recovered from failure."""
+    pass
+```
+
+
+As we mentioned the return value of `schedule` will be `Future` objects if `function` has return values. The `Future` works as a container and will be later-binded with states of either success or complete failure. Overall, this `Future` class has the following benefits:
+
+
+
+*   It allows `schedule` method return immediately after pushing functions to its scheduling queue. It allows these methods to return without needing to wait for acknowledgement from workers.
+*   It serves as the container for values or errors. It would be binded with a value or an error later. When it is rebuilt, we can replace its underlying value silently.
+*   When being passed to `local_result`, we flag it to indicate that this value needs to be fetched eagerly.
+*   (Future work) It captures the lineage between functions and return values so that we can rebuild any poisoned objects. 
+
+```python
+class Future(object):
+
+  def __init__(self, closure):
+    pass
+
+  def _set_value(self, value):
+    pass
+
+  def _set_error(self, error):
+    pass
+
+  def _set_eagerly_fetch(self):
+    pass
+  
+  def _fetch(self):
+    pass
+```
+
+
+
+We can potentially merge this `Future` class with our `Tensor` class.
+
```

Comment:
Note that if this is returned by a public API, we generally require that the returned type is public, or some interface to it. I don't love the idea of a generically named "Future" type in TF, especially because it is suspiciously like the two-stage-Session-running of v1, so we should consider what the contract is here and what the user needs to know about this.

Reply:
You can compare the future to `PerReplica` value in `MirroredStrategy`. In `MirroredStrategy`, `run` returns an opaque type which is `PerReplica` and users can only get its concrete values through `local_results`. Similarly, our `schedule` method returns `Future` (or `PerReplica` of `Future`s when we support multi-GPU), users can only get its concrete values through `local_results`

Our `Tensor` object in async eager is a future-like object as well. It doesn't have any concrete value unless we materialize it via `numpy`. Therefore, we can think of merging the logic to `Tensor` class in the future.

Reply:
With this, it sounds to me the contract is "one would need to use `local_results` to deal with whatever they get from `run()` or `schedule()`", and doesn't mention the type of it. 

Better yet, if possible, we encourage user not to rely on the return value of `schedule()`. For many things that are involved in training such as loss or metrics, they can simply be variables that get updated within the remote function call. User simply reads those values from the client.